api/client/commands.go

```diff
@@ -1674,60 +1697,86 @@ func (cli *DockerCli) CmdPs(args ...string) error {
 		return ss
 	}
 
-	for _, out := range outs.Data {
-		outID := out.Get("Id")
-
-		if !*noTrunc {
-			outID = common.TruncateID(outID)
-		}
-
-		if *quiet {
-			fmt.Fprintln(w, outID)
-
-			continue
-		}
-
-		var (
-			outNames   = stripNamePrefix(out.GetList("Names"))
-			outCommand = strconv.Quote(out.Get("Command"))
-			ports      = engine.NewTable("", 0)
-		)
+	type containerMeta struct {
+		c string
+		i string
+		m string
+		t string
+		s string
+		p string
+		n string
+		z string
+	}
 
+	outp := make([]containerMeta, 0)
+	for _, out := range outs.Data {
+		portTable := engine.NewTable("", 0)
+		portTable.ReadListFrom([]byte(out.Get("Ports")))
+
+		next := containerMeta{
+			c: out.Get("Id"),
+			n: "",
+			m: strconv.Quote(out.Get("Command")),
+			i: out.Get("Image"),
+			t: units.HumanDuration(time.Now().UTC().Sub(time.Unix(out.GetInt64("Created"), 0))) + " ago",
+			s: out.Get("Status"),
+			p: api.DisplayablePorts(portTable),
+			z: fmt.Sprintf("%s", units.HumanSize(float64(out.GetInt64("SizeRw")))),
+		}
+
+		// handle truncation
+		outNames := stripNamePrefix(out.GetList("Names"))
 		if !*noTrunc {
-			outCommand = utils.Trunc(outCommand, 20)
-
+			next.c = common.TruncateID(next.c)
+			next.m = utils.Trunc(next.m, 20)
 			// only display the default name for the container with notrunc is passed
 			for _, name := range outNames {
 				if len(strings.Split(name, "/")) == 1 {
 					outNames = []string{name}
-
 					break
 				}
 			}
 		}
+		next.n = strings.Join(outNames, ",")
 
-		ports.ReadListFrom([]byte(out.Get("Ports")))
+		if next.i == "" {
+			next.i = "<no image>"
+		}
 
-		image := out.Get("Image")
-		if image == "" {
-			image = "<no image>"
+		// handle rootfs sizing
+		if out.GetInt("SizeRootFs") > 0 {
+			next.z = next.z + fmt.Sprintf(" (virtual %s)", units.HumanSize(float64(out.GetInt64("SizeRootFs"))))
 		}
 
-		fmt.Fprintf(w, "%s\t%s\t%s\t%s ago\t%s\t%s\t%s\t", outID, image, outCommand,
-			units.HumanDuration(time.Now().UTC().Sub(time.Unix(out.GetInt64("Created"), 0))),
-			out.Get("Status"), api.DisplayablePorts(ports), strings.Join(outNames, ","))
+		outp = append(outp, next)
+	}
 
-		if *size {
-			if out.GetInt("SizeRootFs") > 0 {
-				fmt.Fprintf(w, "%s (virtual %s)\n", units.HumanSize(float64(out.GetInt64("SizeRw"))), units.HumanSize(float64(out.GetInt64("SizeRootFs"))))
-			} else {
-				fmt.Fprintf(w, "%s\n", units.HumanSize(float64(out.GetInt64("SizeRw"))))
-			}
+	for _, out := range outp {
```

Comment:
Just so down below, I can do the simple fprintf + strings.Join ("\t").. otherwise I have to track if I'm at the last letter, add prints of \t outside the switch, etc. Looked a lot less nicer. Considering this is per line, the memory overhead seemed minimal.